### PR TITLE
[INTEL MKL] Enable DNNL BatchMatMul support with broadcast and update oneDNN to v1.6.4.

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR enables new features from DNNL to perform BatchMatMul with broadcast. Current support is up to rank 12 tensors. Earlier, the BatchMatMul op was using cblas api from Intel Math Kernel Library(MKL).

### CODE REVIEW

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc [hunk 1]
```diff
@@ -149,136 +136,96 @@ class BatchMatMulMkl : public OpKernel {
...
+    const auto ndims_out = out_shape.dims();
+    auto lhs_dims = TFShapeToMklDnnDims(lhs_shape);
+    auto rhs_dims = TFShapeToMklDnnDims(rhs_shape);
+    auto out_dims = TFShapeToMklDnnDims(out_shape);
+
+    // DNNL matmul_primitive requires ranks of inputs and output to be same.
+    // Create dnnl::memory::dims for inputs and output of same rank.
+    // It is assumed here that MatMulBCast object creates output_batch_shape as
+    // a conforming superset of input batch shapes, i.e., ndims_out >=
+    // ndims_lhs and ndims_out >= ndims_lhs.
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc:191]: Typo:
```suggestion
    // ndims_lhs and ndims_out >= ndims_rhs.
```

File: tensorflow/workspace.bzl [hunk 2]
```diff
@@ -209,11 +209,11 @@ def tf_repositories(path_prefix = "", tf_repo_name = ""):
...
         build_file = clean_dep("//third_party/mkl_dnn:mkldnn_v1.BUILD"),
-        sha256 = "aef4d2a726f76f5b98902491a1a4ac69954039aa8e5a1d67ef6ce58ed00e23a6",
-        strip_prefix = "oneDNN-1.5.1",
+        sha256 = "5369f7b2f0b52b40890da50c0632c3a5d1082d98325d0f2bff125d19d0dcaa1d",
+        strip_prefix = "oneDNN-1.6.4",
         urls = [
-            "https://storage.googleapis.com/mirror.tensorflow.org/github.com/oneapi-src/oneDNN/archive/v1.5.1.tar.gz",
-            "https://github.com/oneapi-src/oneDNN/archive/v1.5.1.tar.gz",
+            "https://storage.googleapis.com/mirror.tensorflow.org/github.com/oneapi-src/oneDNN/archive/v1.6.4.tar.gz",
+            "https://github.com/oneapi-src/oneDNN/archive/v1.6.4.tar.gz",
```
Comment (penpornk) [inline tensorflow/workspace.bzl:216]: (No action needed.)
In the future, please do the version upgrade in a separate PR if possible.

**Review by penpornk — CHANGES_REQUESTED** [review]
This is great! Thank you for the PR!

**Comment by mdfaijul** [comment]
@penpornk Thanks for the review and great suggestions. I have addressed them. Please check.

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc [hunk 3]
```diff
@@ -15,27 +15,16 @@ limitations under the License.
...
-// supported by oneDNN (ex. Batchmatmul with broadcasting) we fall back to the
-// default CPU implementation.
-// if INTEL_MKL_DNN_ONLY is not defined, both oneDNN and MKL CBLAS
-// implementations are used. This is only temporary, once we are able handle all
-// cases with oneDNN, CBLAS calls will be removed.
+// This file uses oneDNN library for acceleration of Batch Matrix-Matrix
+// Multiplication (MatMul) operations. We currently register this kernel only
+// for oneDNN supported data types (float, bfloat16). The maximum number of
+// dimensions (rank) for output tensor is 12 in oneDNN. If output tensor rank
+// exceeds 12, we fallback to Eigen library based kernel.
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc [hunk 4]
```diff
@@ -137,6 +118,12 @@ class BatchMatMulMkl : public OpKernel {
 
     out_shape.AddDim(lhs_rows);
     out_shape.AddDim(rhs_cols);
+    // The maximum number of dimensions for a tensor in DNNL is 12.
+    OP_REQUIRES(ctx, out_shape.dims() <= 12,
+                errors::InvalidArgument(
+                    "Rank of output tensor is required as <= 12, ", "but is ",
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc]: Done

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc [hunk 5]
```diff
@@ -149,136 +136,96 @@ class BatchMatMulMkl : public OpKernel {
...
-    std::vector<float> alpha_Array(group_size[0], 1.0);
-    std::vector<float> beta_Array(group_size[0], 0.0);
-    dnnl_gemm_batch<float>(TransA_Array, TransB_Array, M_Array, N_Array,
-                           K_Array, alpha_Array, *A_Array, *B_Array, beta_Array,
-                           *C_Array, group_count, group_size, ctx);
-#endif  // !INTEL_MKL_DNN_ONLY
+  using dims = dnnl::memory::dims;
+
+  // This method makes the rank (ndims) of input same as the output by creating
+  // new axes to the input. For example, if input shape is [a, b, c, d] and
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc:160]: Done

File: tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc]: Done

**Review by penpornk — APPROVED** [review]
Thank you for the quick change!

**Pull request merged**